File: common/jwt-core/src/main/java/com/halcyon/tinder/jwtcore/TokenRevocationService.java
package com.halcyon.tinder.jwtcore;

import com.halcyon.tinder.rediscache.CacheManager;
import java.time.Duration;
import lombok.RequiredArgsConstructor;
import org.springframework.context.annotation.ComponentScan;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
@ComponentScan(basePackages = {
        "com.halcyon.tinder.jwtcore",
        "com.halcyon.tinder.rediscache"
})
public class TokenRevocationService {

    private final CacheManager cacheManager;
    private final JwtProvider jwtProvider;

    private static final String TOKEN_PREFIX = "revoked-token:";

    public void revokeToken(String jwtToken) {
        long expiryInSeconds = jwtProvider.getRemainingTokenLifetime(jwtToken);
        String key = TOKEN_PREFIX + jwtToken;

        cacheManager.save(key, Duration.ofSeconds(expiryInSeconds));
    }

    public boolean isRevoked(String jwtToken) {
        String key = TOKEN_PREFIX + jwtToken;
        return cacheManager.isPresent(key);
    }
}

File: common/jwt-core/src/main/kotlin/com/halcyon/tinder/jwtcore/TokenRevocationService.kt
package com.halcyon.tinder.jwtcore

import com.halcyon.tinder.rediscache.CacheManager
import org.springframework.stereotype.Service
import java.time.Duration

@Service
class TokenRevocationService(
    private val cacheManager: CacheManager,
    private val jwtProvider: JwtProvider
) {

    companion object {
        private const val TOKEN_PREFIX = "revoked-token:"
    }

    fun revokeToken(jwtToken: String) {
        val expiryInSeconds = jwtProvider.getRemainingTokenLifetime(jwtToken)
        val key = TOKEN_PREFIX + jwtToken

        cacheManager.save(key, Duration.ofSeconds(expiryInSeconds))
    }

    fun isRevoked(jwtToken: String): Boolean {
        val key = TOKEN_PREFIX + jwtToken
        return cacheManager.isPresent(key)
    }
}